### PR TITLE
Feature/minimization

### PR DESCRIPTION
**Pull request by MaxOstrowski**
Got inspired to try exponential decrease minimization.
Do you think this is useful ?

### CODE REVIEW

**Comment by MaxOstrowski** [comment]
Sample program:
```
{opt(1..100)}.
&diff{0 - step} <= -O :- opt(O).
&diff{step - opt} <= 0.
#heuristic opt(N) : N=1..100. [1,true]
&diff{0-opt} <= 0.
```

`clingo-dl temp.lp --minimize-variable=opt --heurist=domain --minimize-strategy=exponential --stats=2`
`clingo-dl temp.lp --minimize-variable=opt --heurist=domain --minimize-strategy=linear --stats=2`

I do not know how to correctly check for overflows (as numbers are growing fast) for the Clingo::Numbers, any suggestions ?

**Comment by MaxOstrowski** [comment]
I still do see the
`INTERRUPTED  : 1`
and I do not where this comes from...

**Comment by MaxOstrowski** [comment]
Fixed local imbui on command line parsing.
Still not don't what to do with your idea of `n*(f^i)`.

**Comment by MaxOstrowski** [comment]
Any plan how to continue this ?

**Comment by rkaminsk** [comment]
I don't have time to look into this atm. I also have no project where I could test it. Maybe we just keep it around as a branch.

**Comment by rkaminsk** [comment]
I could not really make sense of your code so I rewrote it into something that is hopefully easier to read. I changed (at least) the following points:
- Externals are only introduced if required (I think this is important for performance).
- I separated the optimization logic into a dedicated class (the previous code was a chaotic mess).
- I added some trivial test cases (maybe you can add some more).

For an adjustment factor of `1` it should behave as before. Please have a look if it does what it should.

File: libclingo-dl/clingo-dl/app.hh [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+                "&diff { s-0 } <= b."
+                // add a retractable bound
+                "#program __sb(s,b)."
+                "#external __sb(b). [true]"
+                "&diff { s-0 } <= b :- __sb(b).");
+        });
+        if (opt_cfg_.has_initial) {
+            upper_bound_ = opt_cfg_.initial;
+        }
+        for (int i = 0; i < 100; ++i) {
```
Comment (MaxOstrowski) [inline libclingo-dl/clingo-dl/app.hh:132]: shouldnt this be infinite ? Or the number of models requested ?

**Review by MaxOstrowski — COMMENTED** [review]
Looks awesome.
I found this magic constant in your code.
Also, if the variable is not bound we easily get an "integer overflow" error, which is nice.
`&diff{a - 0} <= 5.`
`clingo-dl --minimize-variable=a --minimize-factor=2`
```
Answer: 1
dl(a,-2147483647) __sb(-2147483647)
*** ERROR: (clingo-dl): integer overflow
```

Also it might be useful to hide __sb ?

File: libclingo-dl/clingo-dl/app.hh (hunk 1 above)
Comment (rkaminsk) [reply libclingo-dl/clingo-dl/app.hh:132]: This is a leftover from testing. I'll fix it. Thanks.

**Comment by rkaminsk** [comment]
> Also it might be useful to hide __sb ?

There is no way to hide it.

**Comment by rkaminsk** [comment]
I forgot to mention that I did not take over the code to add lower bounds. If I remember correctly, at least with normal minimization, such kind of constraints do not help.

**Pull request merged**